File: write_dry_run.py
from pathlib import Path

def write_dry_run(file_path, content, dry_run, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if dry_run:
        return f"[Dry Run] Would write to {file_path}:\n{content}"
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(content)
    return f"Wrote content to {file_path}"}
```

File: packages/core/src/tools/write-dry-run.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { promises as fs } from 'fs';
import { resolve } from 'path';
import { checkFilePermission } from '../services/filePermissionService.js';

interface DryRunConfig {
  dryRun?: boolean;
}

export async function writeDryRun(
  filePath: string,
  content: string,
  dryRun: boolean,
  config: DryRunConfig,
): Promise<string> {
  if (!(await checkFilePermission(filePath, 'write', config))) {
    throw new Error(`Write access denied for ${filePath}`);
  }

  const resolvedPath = resolve(filePath);

  if (dryRun) {
    return `[Dry Run] Would write to ${resolvedPath}:\n${content}`;
  }

  await fs.writeFile(resolvedPath, content, 'utf-8');
  return `Wrote content to ${resolvedPath}`;
}
